File: solutions/julia/perfect-numbers/1/perfect-numbers.jl
function aliquot_sum(num)
	num <= 0 && throw(DomainError(""))

    a, b, divisors = 1, num, Int[]
    while b > a
        mod(b, 1) == 0 && push!(divisors, a, b)
        a += 1
        b = num / a
    end
    sum(divisors) - num
end

isperfect(num) = num == aliquot_sum(num)
isabundant(num) = num < aliquot_sum(num)
isdeficient(num) = num > aliquot_sum(num)

File: solutions/csharp/perfect-numbers/1/PerfectNumbers.cs
using System;

public static class PerfectNumbers
{
    public static Classification Classify(int number) => AliquotSum(number) switch
    {
        1 => Classification.Deficient,
        var s when s < number => Classification.Deficient,
        var s when s > number => Classification.Abundant,
        _ => Classification.Perfect
    };

    private static int AliquotSum(int num)
    {
        if (num <= 0)
            throw new ArgumentOutOfRangeException(nameof(num), "Non positive number!");

        int sum = 1, a = 2;
        var b = num / 2d;

        while (a < b)
        {
            if (b % 1 == 0)
                sum += a + (int)b;

            b = (double)num / ++a;
        }

        return sum;
    }
}

public enum Classification
{
    Perfect,
    Abundant,
    Deficient
}